### ISSUE TITLE
Pack array into lower triangular half of matrix using native TensorFlow ops.

### ISSUE THREAD

**Issue by alexggmatthews** [GPflow/GPflow#439]
At the moment we have our own user ops to do this which was necessary at the time.

But this is probably now possible in atleast one way:

https://stackoverflow.com/questions/37307519/packing-array-into-lower-triangular-of-a-tensor

Although as @markvdw point's out:

https://github.com/GPflow/GPflow/issues/423

it will need to be profiled. 

Although the user ops have been very helpful it would decrease maintenance overhead and ease installation to have only native TF ops. It would also help us become a pip package:

https://github.com/GPflow/GPflow/pull/381

Questions:

1) What is the best way to do this?
2) Is it acceptably fast?

**Comment by alexggmatthews** [comment]
@jameshensman I'm going to have a look at this issue this morning.

**Comment by alexggmatthews** [comment]
Question 3:) Is there a way to do in that already has GPU ops?

**Comment by alexggmatthews** [comment]
Looks like it is going to be better to use scatter_nd because it has more support at Google generally:

https://github.com/tensorflow/tensorflow/issues/6391

And in particular has a GPU implementation.

**Comment by alexggmatthews** [comment]
I am working on this here:

https://github.com/GPflow/GPflow/tree/native_tf_triang

**Comment by alexggmatthews** [comment]
Maybe we should switch to matrix_band_part while we're at it @jameshensman ? Did you say diag_part didn't have a GPU op?

**Comment by alexggmatthews** [comment]
Or was it diag?

**Comment by jameshensman** [comment]
diag and diagpart do not have gpu ops if I remember correctly.

**Comment by alexggmatthews** [comment]
Should be fixed my matrix_diag and matrix_diagpart but let's do that in a separate PR. Unimpressed with that redundancy in tf core.

**Comment by jameshensman** [comment]
They do have slightly different functionality iof the matrix has more than 2 dimensions.

**Comment by alexggmatthews** [comment]
In terms of the vec_to_triang op . It currently works fine apart from the shape inference. I.e I have a (DxM) -> (DxNxN) but the mapping from M -> N and the associated checks need to be done in Tensor shape space and look horrible at the moment.

**Comment by alexggmatthews** [comment]
In fact I think the approach I'm using will only work when we can specify the shape of the matrix at graph creation time. We need an explicit value for N so that we can call np.tril_indeces which is then passed to scatter_nd. Since np.tril_indeces is not a TensorFlow function we can't pass it a tf shape @jameshensman any thoughts?

**Comment by jameshensman** [comment]
Two thoughts from me:

1) you could vreate a new python tensorflow op that wraps np.tril_indices
2) The lower triangular transform currently requires num_matrices (your D). Would it be impossible to simply specify D and N at creation time, and make them tf.constants?

**Comment by javdrher** [comment]
I haven't been following this in detail, but might a tf.tril_indices be useful so you can generate the indices in the graph?
Weakly tested attempt:
```python
import tensorflow as tf
import numpy as np

def tf_tril_indices(N, k=0):
   M1 = tf.tile(tf.expand_dims(tf.range(N), axis=0), [N,1])
   M2 = tf.tile(tf.expand_dims(tf.range(N), axis=1), [1,N])
   mask = (M1-M2) >= -k
   ix1 = tf.boolean_mask(M2, tf.transpose(mask))
   ix2 = tf.boolean_mask(M1, tf.transpose(mask))
   return ix1, ix2

with tf.Session() as sess:
    X = tf.ones([4, 4])
    
    # Standard
    ix1, ix2 = tf_tril_indices(tf.shape(X)[0])
    sess.run(tf.global_variables_initializer())
    print (sess.run([ix1, ix2]))
    print( np.tril_indices(4) )
    
    # With diagonal offset
    ix1, ix2 = tf_tril_indices(tf.shape(X)[0], 1)
    sess.run(tf.global_variables_initializer())
    print (sess.run([ix1, ix2]))
    print( np.tril_indices(4, 1) )
```

**Comment by rhofour** [comment]
Is there already some benchmarking code for this?

**Comment by alexggmatthews** [comment]
@javdrher Nice! looks to working OK with the current variant but may keep this one in mind if we have any probs.

**Comment by alexggmatthews** [comment]
Fixed in #440
